lib/elixir/lib/gen_server.ex

```diff
@@ -748,7 +748,16 @@ defmodule GenServer do
   """
   @spec stop(server, reason :: term, timeout) :: :ok
   def stop(server, reason \\ :normal, timeout \\ :infinity) do
-    :gen.stop(server, reason, timeout)
+    case whereis(server) do
+      nil ->
+        exit(:noproc)
+
+      pid when pid == self() ->
+        exit({:stopping_self, {__MODULE__, :stop, [server, reason, timeout]}})
```

Comment:
Let's use `:calling_self` because that already provides us a good error message.